### ISSUE TITLE
Phase 1: Cross-Encoder Reranker

### ISSUE THREAD

**Issue by tinkermonkey** [tinkermonkey/context-library#153]
Implement the `Reranker` class in `src/context_library/retrieval/reranker.py` to reorder vector search candidates using a cross-encoder model, improving retrieval quality beyond bi-encoder similarity scores.

## Requirements
- Implement a `Reranker` class that wraps `sentence_transformers.CrossEncoder` for relevance scoring
- Accept `query: str` and `candidates: list[RetrievalResult]` and return a reordered `list[RetrievalResult]`
- Support an optional `top_k: int | None` parameter to truncate results after reranking
- Use the `content` field already available on `RetrievalResult.chunk.content` — no SQLite round-trips
- Replace `similarity_score` on each returned `RetrievalResult` with the normalized cross-encoder score (in range [0, 1])
- Return an empty list when `candidates` is empty
- Handle `top_k=None` by returning all reranked candidates
- Default model: `cross-encoder/ms-marco-MiniLM-L-6-v2`
- Support batch inference: all candidates scored in a single `CrossEncoder.predict()` call
- No new pip dependencies required — `sentence-transformers` already includes `CrossEncoder`
- The `Reranker` class should NOT be stateless (it holds the `CrossEncoder` instance); model load happens once at construction
- Reranker is optional post-processing; callers that don't need it continue to call `retrieve()` directly
- Add tests in `tests/retrieval/test_reranker.py`

## Design Guidance
**File:** `src/context_library/retrieval/reranker.py`

**Class signature:**
```python
from sentence_transformers import CrossEncoder
from context_library.retrieval.query import RetrievalResult

class Reranker:
    def __init__(self, model_name: str = "cross-encoder/ms-marco-MiniLM-L-6-v2") -> None:
        self._model_name = model_name
        self._model = CrossEncoder(model_name)

    def rerank(
        self,
        query: str,
        candidates: list[RetrievalResult],
        top_k: int | None = None,
    ) -> list[RetrievalResult]: ...
```

**Score normalization:** Cross-encoder scores are raw logits (unbounded). Normalize via sigmoid:
```python
import math
def _sigmoid(x: float) -> float:
    return 1.0 / (1.0 + math.exp(-x))
```
Apply sigmoid to each raw score, then reconstruct `RetrievalResult` with the normalized score as `similarity_score` (range [0, 1], matching the field validator at `query.py:40-48`).

**Batch inference pattern:**
```python
pairs = [(query, r.chunk.content) for r in candidates]
raw_scores = self._model.predict(pairs)  # single forward pass
```

**Mirrors `Embedder` design (`src/context_library/core/embedder.py:8`):** class holds model instance loaded at construction time.

**`RetrievalResult` is frozen (`ConfigDict(frozen=True)` at `query.py:34`).** Reconstruct new instances:
```python
from pydantic import BaseModel
return RetrievalResult(
    chunk=candidate.chunk,
    lineage=candidate.lineage,
    similarity_score=normalized_score,
)
```

**Test patterns to follow (from `tests/retrieval/test_query.py`):**
```python
from unittest.mock import MagicMock, patch

def _make_hash(char: str) -> str:
    return char * 64

def _create_test_chunk(hash_char, chunk_index=0):
    return Chunk(
        chunk_hash=_make_hash(hash_char),
        content=f"Test chunk {hash_char}",
        chunk_index=chunk_index,
        chunk_type="standard",
    )

def _create_test_lineage(hash_char, source_id="source_1"):
    return LineageRecord(
        chunk_hash=_make_hash(hash_char),
        source_id=source_id,
        source_version_id=1,
        adapter_id="test-adapter",
        domain=Domain.NOTES,
        normalizer_version="1.0.0",
        embedding_model_id="all-MiniLM-L6-v2",
    )
```

Mock `CrossEncoder.predict()` in unit tests to avoid model downloads:
```python
with patch("context_library.retrieval.reranker.CrossEncoder") as mock_ce:
    mock_instance = MagicMock()
    mock_ce.return_value = mock_instance
    mock_instance.predict.return_value = [2.0, -1.0, 0.5]
    reranker = Reranker()
    results = reranker.rerank("query", candidates)
```

## Acceptance Criteria
- [ ] `Reranker.__init__` loads a `CrossEncoder` model at construction time
- [ ] `reranker.rerank(query, candidates)` returns all candidates reordered by descending cross-encoder score
- [ ] `reranker.rerank(query, candidates, top_k=3)` returns at most 3 results
- [ ] Each returned `RetrievalResult.similarity_score` is in range [0, 1] (sigmoid-normalized)
- [ ] `reranker.rerank(query, [])` returns `[]` without error
- [ ] All candidates are scored in a single `CrossEncoder.predict()` call (batch inference)
- [ ] Unit tests mock `CrossEncoder.predict()` — no real model download required in CI
- [ ] Tests verify score ordering, top_k truncation, score normalization, and empty input handling
- [ ] `tests/retrieval/test_reranker.py` contains the complete test suite
- [ ] Code reviewed and approved

## Parent Issue
Part of #151

## Discussion
This work is detailed in discussion [152](https://github.com/tinkermonkey/context-library/discussions/152)

**Comment by tinkermonkey** [comment]
# Code Review

---



I'll start by reviewing the previous feedback, examining the current code changes, and checking CLAUDE.md for any specialized agents.Let me examine the actual file diffs for the files that are in scope for issue #153 (the reranker), and check what was there at the base commit.Now let me check the other files that were changed to see if they're in scope for this review (issue #153 is the reranker, but there are also issues #154-156). Let me also look at the existing codebase patterns for context.Now let me check if there's previous review feedback to reference, and also look at the other changed files to understand scope.Now let me check if there are other files that were changed that are outside the reranker scope. The issue description is specifically about the reranker, so let me verify whether the other files (provenance, document_store, models, vector_store) are part of separate issues (#154-156) and thus out of scope for this review.Good. So the reranker files were only changed in commit `a1b943f` (issue #153). The other files are from separate issues (#154-156) and are out of scope. Now let me check the `top_k` type annotation - the requirements say `int | None` but the code uses `Optional[int]`. Let me also verify the design guidance spec compliance point by point.Now I have a thorough understanding of the code. Let me also check if there's a `.claude_prompt_4e0b93f9-643b-47a4-8822-3d75af12c5bd.txt` file that might contain the previous review feedback.This is iteration 2, meaning I need to reference the previous review's issues. Since this is the first time I'm reviewing in this conversation, and the previous review is not included in my context, I need to assess the code fresh but based on the stated requirements. Let me verify the test helper functions match the project patterns.Good - the test helpers in `test_reranker.py` match the pattern used in `test_query.py` exactly (including `context_header` and `domain_metadata` fields).

Now I have all the information I need to write the review. Let me verify all acceptance criteria against the implementation:

1. `Reranker.__init__` loads CrossEncoder at construction - YES (`reranker.py:59-60`)
2. `rerank()` returns all candidates reordered by descending score - YES (`reranker.py:116`)
3. `top_k=3` returns at most 3 - YES (`reranker.py:119-120`)
4. Scores in [0,1] via sigmoid - YES (`reranker.py:106`, `_sigmoid` function)
5. Empty candidates returns `[]` - YES (`reranker.py:96-97`)
6. Single `predict()` call - YES (`reranker.py:100-101`)
7. Tests mock `CrossEncoder.predict()` - YES (all tests use `@patch`)
8. Tests verify ordering, top_k, normalization, empty input - YES
9. Tests in `tests/retrieval/test_reranker.py` - YES

The code matches requirements and design guidance closely. The implementation is clean, follows the existing `Embedder` pattern, and tests are comprehensive. Since this is a re-review iteration 2 and I don't have the explicit previous feedback in context, I'll evaluate the current state comprehensively.

### Status
**APPROVED**

### Previous Issues Status

Since the previous review feedback is not included in this iteration's context, I evaluated the full implementation against all requirements and acceptance criteria. The code in its current state satisfies all requirements — any issues from the prior review appear to have been resolved.

### New Issues Found

#### Critical (Must Fix)
None

#### High Priority (Should Fix)
None

#### Advisory (Out of Scope / FYI)
- **`Optional[int]` vs `int | None`**: The design guidance specifies `top_k: int | None = None` but the implementation uses `Optional[int]`. Both are equivalent at runtime and `Optional` is consistent with the rest of the codebase (e.g., `query.py:10` imports `Optional`). No change needed.
- **Verbose docstring on `_sigmoid`**: The `_sigmoid` function has a 7-line docstring for a 5-line function. This is cosmetic and consistent with the project's existing documentation style in `embedder.py`, so no change needed.

### Summary

The `Reranker` implementation is clean, correct, and well-tested. It faithfully follows the design guidance: class signature matches, `_sigmoid` normalization is numerically stable, batch inference uses a single `predict()` call, frozen `RetrievalResult` objects are properly reconstructed, and the class mirrors the `Embedder` pattern. The test suite covers all acceptance criteria (ordering, top_k truncation, score normalization, empty input, batch inference verification, immutability) with properly mocked `CrossEncoder`. All 9 acceptance criteria are met.

---
_Generated by Orchestrator Bot 🤖_
_Processed by the code_reviewer agent_
